Mathlib/RingTheory/PowerSeries/Substitution.lean

```diff
@@ -0,0 +1,381 @@
+    map_X]
+  erw [AlgHom.comp_apply]
+  simp only [Polynomial.aeval_X, MvPolynomial.coeToMvPowerSeries.algHom_apply,
+    Algebra.id.map_eq_id, MvPowerSeries.map_id, MvPolynomial.coe_X, RingHom.id_apply]
+  rfl
+
+theorem substAlgHom_coe (ha : HasSubst a) (p : Polynomial R) :
+    substAlgHom ha (p : PowerSeries R) = ↑(Polynomial.aeval a p) := by
+  rw [p.toPowerSeries_toMvPowerSeries, substAlgHom]
+  erw [MvPowerSeries.coe_substAlgHom]
+  rw [MvPowerSeries.subst_coe, ← AlgHom.comp_apply]
+  apply AlgHom.congr_fun
+  apply Polynomial.algHom_ext
+  simp only [AlgHom.coe_comp, Function.comp_apply, Polynomial.aeval_X, MvPolynomial.aeval_X]
+
+theorem substAlgHom_X (ha : HasSubst a) :
+    substAlgHom ha (X : R⟦X⟧) = a := by
+  rw [← Polynomial.coe_X, substAlgHom_coe, Polynomial.aeval_X]
+
+theorem subst_coe (ha : HasSubst a) (p : Polynomial R) :
+    subst (R := R) a (p : PowerSeries R) = ↑(Polynomial.aeval a p) := by
+  rw [← coe_substAlgHom ha, substAlgHom_coe]
+
+theorem subst_X (ha : HasSubst a) :
+    subst a (X : R⟦X⟧) = a := by
+  rw [← coe_substAlgHom ha, substAlgHom_X]
+
+theorem HasSubst.comp {a : PowerSeries S} (ha : HasSubst a)
+    {b : MvPowerSeries υ T} (hb : HasSubst b):
+    HasSubst (substAlgHom hb a) :=
+  MvPowerSeries.IsNilpotent_subst hb.const ha
+
+variable
+    {υ : Type*} -- [DecidableEq υ]
+    {T : Type*} [CommRing T] [Algebra R T]
+    {a : PowerSeries S}
+    {b : MvPowerSeries υ T}
+    {a' : MvPowerSeries τ S}
+    {b' : τ → MvPowerSeries υ T}
+
+theorem substAlgHom_comp_substAlgHom [Algebra S T] [IsScalarTower R S T]
+    (ha : HasSubst a) (hb : HasSubst b) :
+    ((substAlgHom hb).restrictScalars R).comp (substAlgHom ha)
+      = substAlgHom (ha.comp hb) :=
+  MvPowerSeries.substAlgHom_comp_substAlgHom _ _
+
+theorem substAlgHom_comp_substAlgHom_apply [Algebra S T] [IsScalarTower R S T]
+    (ha : HasSubst a) (hb : HasSubst b) (f : PowerSeries R) :
+    (substAlgHom hb) (substAlgHom  ha f) = substAlgHom (ha.comp hb) f :=
+  DFunLike.congr_fun (substAlgHom_comp_substAlgHom ha hb) f
+
+theorem subst_comp_subst [Algebra S T] [IsScalarTower R S T]
+    (ha : HasSubst a) (hb : HasSubst b) :
+    (subst b) ∘ (subst a) = subst (R := R) (subst b a) := by
+  simpa only [funext_iff, DFunLike.ext_iff, AlgHom.coe_comp, AlgHom.coe_restrictScalars',
+    Function.comp_apply, coe_substAlgHom]
+    using substAlgHom_comp_substAlgHom (R := R) ha hb
+
+theorem subst_comp_subst_apply [Algebra S T] [IsScalarTower R S T]
+    (ha : HasSubst a) (hb : HasSubst b) (f : PowerSeries R) :
+    subst b (subst a f) = subst (subst b a) f :=
+  congr_fun (subst_comp_subst (R := R) ha hb) f
+
+theorem _root_.MvPowerSeries.rescaleUnit (a : R) (f : R⟦X⟧) :
+    MvPowerSeries.rescale (Function.const _ a) f = rescale a f := by
+  ext d
+  rw [coeff_rescale, coeff, MvPowerSeries.coeff_rescale]
+  simp [smul_eq_mul, Finsupp.prod_single_index]
+
+namespace Example
```

Comment:
In my opinion we can remove the example entirely. If you want we keep certain results, like `PowerSeries.order_pow`, but I don't think we want a a by hand definition of the exponential (I mean, it already exists in various degree of generality in mathlib).

Reply:
It was there as a show case for @b-mehta , I plan to delete it.

Reply:
The result for `order_pow` is in PR #23993